### PR TITLE
**Add password to rootCert.pfx. and more...

### PR DESCRIPTION
**Pull request by mohammadlachgar**
**Save all fake certificates in folder "crts"(will be save in proxy dll directory)
 for can load the certificate and not make new certificate every time .
**Manually load a RootCertificate...

Doneness:
- [*] Build is okay - I made sure that this change is building successfully.
- [*] No Bugs - I made sure that this change is working properly as expected. It doesn't have any bugs that you are aware of. 
- [ ] Branching - If this is not a hotfix, I am making this request against develop branch

### CODE REVIEW

File: Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs [hunk 1]
```diff
@@ -146,7 +149,7 @@ public X509Certificate2 MakeCertificate(string sSubjectCn, bool isRoot, X509Cert
             {
                 try
                 {
-                    x509Certificate.FriendlyName = subjectName;
+                    x509Certificate.FriendlyName = System.Text.RegularExpressions.Regex.Replace(subjectName.ToLower(), @"^" + "CN".ToLower() + @"\s*=\s*", "");
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs]: Why don't you write simply 
@"^cn\s*=\s*"
?

And this regex sshould be compiled.

File: Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs [hunk 2]
```diff
@@ -199,11 +202,17 @@ private static X509Certificate2 WithPrivateKey(Org.BouncyCastle.X509.X509Certifi
...
+
                 var kp = DotNetUtilities.GetKeyPair(signingCertificate.PrivateKey);
+
                 return GenerateCertificate(hostName, subjectName, signingCertificate.Subject, validFrom, validTo, issuerPrivateKey: kp.Private);
+
             }
         }
 
+
+
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs]: Please do not insert new lines.

File: Titanium.Web.Proxy/Network/CertificateManager.cs [hunk 3]
```diff
@@ -73,20 +100,23 @@ internal CertificateEngine Engine
 
         private X509Certificate2 rootCertificate;
 
+        public bool only_load_rootCert { get; set; }
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/CertificateManager.cs]: please follow the guidelines: https://github.com/justcoding121/Titanium-Web-Proxy/wiki

File: Titanium.Web.Proxy/ProxyServer.cs [hunk 4]
```diff
@@ -520,6 +563,18 @@ public void Start()
                 throw new Exception("Proxy is already running.");
             }
 
+
+            try
+            {
+                if (ProxyEndPoints.Any(x => (x as ExplicitProxyEndPoint).GenericCertificate == null))
```
Comment (honfika) [inline Titanium.Web.Proxy/ProxyServer.cs]: This is strange.
1st: when you always access a property you should use direct cast. This will help other developers to know what the problem is. When you use "as", it may return null and you will get a nullreference exception. If you use direct cast, you will receive an invalid cast exception which is better.
2nd: I think you added the try-catch because of the possible nullreference exceptions (cast problems)
Probaby you wanted to write this:
```
                if (ProxyEndPoints.OfType<ExplicitProxyEndPoint>().Any(x => x.GenericCertificate == null))
                {
                    EnsureRootCertificate();
                }

```

**Comment by justcoding121** [comment]
Please also state why this change is required. I think what you are looking for is already supported.

File: Examples/Titanium.Web.Proxy.Examples.Wpf/MainWindow.xaml.cs [hunk 5]
```diff
@@ -88,7 +113,9 @@ public MainWindow()
             proxyServer.TunnelConnectResponse += ProxyServer_TunnelConnectResponse;
             proxyServer.ClientConnectionCountChanged += delegate { Dispatcher.Invoke(() => { ClientConnectionCount = proxyServer.ClientConnectionCount; }); };
             proxyServer.ServerConnectionCountChanged += delegate { Dispatcher.Invoke(() => { ServerConnectionCount = proxyServer.ServerConnectionCount; }); };
-            proxyServer.Start();
+            
+                proxyServer.Start(); 
+
```
Comment (honfika) [inline Examples/Titanium.Web.Proxy.Examples.Wpf/MainWindow.xaml.cs]: Please remove the unnecessary empty lines from everywhere. This comment is only an example.
- The code should not have 2 consecutive empty lines
- Do not put empty lines in the beginning of a block (after "{")

Please fix the indent, `proxyServer.Start()` should be moved left by 4 spaces

File: Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs [hunk 6]
```diff
@@ -46,7 +48,7 @@ internal BCCertificateMaker(Action<Exception> exceptionFunc)
         /// <param name="signingCert">The signing cert.</param>
         /// <returns>X509Certificate2 instance.</returns>
         public X509Certificate2 MakeCertificate(string sSubjectCn, bool isRoot, X509Certificate2 signingCert = null)
-        {
+        {
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs]: Why did you add this space?

File: Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs [hunk 7]
```diff
@@ -145,8 +148,8 @@ public X509Certificate2 MakeCertificate(string sSubjectCn, bool isRoot, X509Cert
             if (!doNotSetFriendlyName)
             {
                 try
-                {
-                    x509Certificate.FriendlyName = subjectName;
+                { 
+                    x509Certificate.FriendlyName = System.Text.RegularExpressions.Regex.Replace(subjectName.ToLower(), @"^CN\s*=\s*", "");
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/Certificate/BCCertificateMaker.cs]: ToLower and the expresion contains upper cased CN? Anyway you should use case insensitive compiled regex.

File: Titanium.Web.Proxy/Network/CertificateManager.cs [hunk 8]
```diff
@@ -110,6 +140,18 @@ internal X509Certificate2 RootCertificate
             }
         }
 
+
+        private string password_rootCert = string.Empty;
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/CertificateManager.cs]: Check the naming conventions. Your code contains a lot of names (field, property, method) which has incorrect names. Please fix all of them.

File: Titanium.Web.Proxy/Network/CertificateManager.cs [hunk 9]
```diff
@@ -110,6 +140,18 @@ internal X509Certificate2 RootCertificate
...
+
+        public string Password_rootCert
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/CertificateManager.cs]: This can be an auto property

File: Titanium.Web.Proxy/Network/CertificateManager.cs [hunk 10]
```diff
@@ -142,49 +185,96 @@ private string GetRootCertificatePath()
...
+                }
+            }
+
+             
 
             try
-            {
-                RootCertificate = CreateCertificate(RootCertificateName, true);
+            { 
+                RootCertificate = CreateCertificate(RootCertificateName, true);
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/CertificateManager.cs]: Spaces added to the end of the lines.

File: Titanium.Web.Proxy/Network/CertificateManager.cs [hunk 11]
```diff
@@ -195,18 +285,71 @@ public bool CreateTrustedRootCertificate(bool persistToFile = true)
...
+        {
+            SetInfo_LoadRootCertificate(fileName, password, overwriteRootCert, StorageFlag);
+             RootCertificate = LoadRootCertificate(); 
+
+            return (RootCertificate != null);
+
+        }
+      
+        private string passwordx = string.Empty;
+        private string filename_loadx = "";
```
Comment (honfika) [inline Titanium.Web.Proxy/Network/CertificateManager.cs]: What is that "x" at the end of the name? please use string.Empty

File: Titanium.Web.Proxy/ProxyServer.cs [hunk 12]
```diff
@@ -520,6 +563,18 @@ public void Start()
...
 
+
+            try {
+                  if (ProxyEndPoints.OfType<ExplicitProxyEndPoint>().Any(x => x.GenericCertificate == null))
+                  {
+                    EnsureRootCertificate();
+                  }
+            }
+            catch (Exception ex)
+            {
```
Comment (honfika) [inline Titanium.Web.Proxy/ProxyServer.cs]: Please remove the try-catch

**Comment by honfika** [comment]
I've merged your changes, could you please test it?
